### DISCUSSION TITLE
How to enable ipv6 stateless auto configuration (SLAAC) (and enable ipv6 forwarding?) for gateways/routers?

### DISCUSSION THREAD

**Discussion by topas-rec** [firewalld/firewalld#1375] [General]
Hello everyone.

I am running an arch router and want to make it ipv6 aware, so that the SLAAC works for the attached network.

Now to enable ipv6 arch has [a nice wiki entry that explains how to enable ipv6 on gateways](https://wiki.archlinux.org/title/IPv6#For_gateways).
I states that

> The gateway must also allow the traffic of ipv6-icmp packets on all basic chains.

As mentioned in the arch wiki I use `radvd` as the software that handles the router messages (advertisements and router solicitation - all that is necessary for SLAAC). When it starts, it warns:

> Aug 13 20:53:20 coreRouter systemd[1]: Started IPv6 Router Advertisement Daemon.
Aug 13 20:53:20 coreRouter radvd[2365]: [Aug 13 20:53:20] radvd (2365): version 2.19 started
Aug 13 20:53:20 coreRouter radvd[2365]: [Aug 13 20:53:20] radvd (2365): IPv6 forwarding setting is: 0, should be 1 or 2
Aug 13 20:53:20 coreRouter radvd[2365]: [Aug 13 20:53:20] radvd (2365): IPv6 forwarding seems to be disabled, but continuing anyway
Aug 13 20:53:20 coreRouter radvd[2365]: [Aug 13 20:53:20] radvd (2365): IPv6 forwarding on interface seems to be disabled, but continuing anyway

Forwarding for ipv6 is indeed currently disabled - not by me intentionally - just the default linux setup I guess. I just skipped the setup for icmp-v6 messages because I didn't find anything on how to enable it in firewalld.

_Side note: I do not get an ipv6 address from `radvd` in its current (warning-about-ipv6-forwarding) state, but it delivers the configured DNS servers and the default route for ipv6. I don't want to hide this information, but it is not relevant I think._

**-> So my general question is** now how do I enable "that ipv6-icmp functionality" in question the right way with firewalld?

My current research:

ipv4 forwarding is enabled by enabling masquerading. NAT is rather unusual in ipv6. So enabling masquerading seems the wrong way.
I found an older issue about that discussed that: https://github.com/firewalld/firewalld/issues/683
It suggests to use masquerading even for ipv6, but before the [introduction of policy-objectes.](https://firewalld.org/2020/09/policy-objects-introduction)
The introduction of policy-objects adds a predefined ipv6-icmp policy - active by default.
I checked this policy and it is active on my system (with the default zones: `ANY` and `HOST`).

I also found [the issue that is related to the introduction of policy-objects](https://github.com/firewalld/firewalld/issues/594#issuecomment-624096417).

**My additional questions are**: Is this ipv6 policy object intended to do what is necessary for ipv6 SLAAC?

I hope this is all clearly explained. If not, please ask.
Tobias

**Comment by erig0** [comment]
> Forwarding for ipv6 is indeed currently disabled - not by me intentionally - just the default linux setup I guess.

Yes. Disabled in Linux default.

> NAT is rather unusual in ipv6. So enabling masquerading seems the wrong way.

Indeed. I suggest you do not use IPv6 NAT. If you do, then you must use a rich rule to allow IPv6 masquerade.

> I just skipped the setup for icmp-v6 messages because I didn't find anything on how to enable it in firewalld.

firewalld allows a subset of ICMPv6 by default. Enough for clients to get IPv6 connectivity.

```
# firewall-cmd --info-policy allow-host-ipv6
allow-host-ipv6 (active)
  priority: -15000
  target: CONTINUE
  ingress-zones: ANY
  egress-zones: HOST
...
  rich rules: 
        rule family="ipv6" icmp-type name="neighbour-advertisement" accept
        rule family="ipv6" icmp-type name="neighbour-solicitation" accept
        rule family="ipv6" icmp-type name="router-advertisement" accept
        rule family="ipv6" icmp-type name="redirect" accept
```

If you want your node to run as an IPv6 _router_, then you'll have to allow [ND IPv6 things](https://en.wikipedia.org/wiki/Neighbor_Discovery_Protocol). Most notably for your use case is **Router Solicitation (Type 133)**.

We could add an `allow-router-ipv6` policy as well. IMO, it would be disabled by default.


# tl;dr

1. Allow Router Solicitations so clients can find your router (radvd).
2. Enable IPv6 forwarding via sysctl.
3. Allow forwarding between zones with a policy (#1376).
  - only necessary if you use multiple zones; forwarding inside the same zone is allowed by default

**Comment by topas-rec** [comment]
I finally came back to this and had since time to think and read about it.
In the end this is exactly how I did it.
Than you for taking the time to help out.

(I'll dive into extra ipv6 router things later)
